shell/packages/sandstorm-accounts-merge/accounts-merge-server.js

```diff
@@ -0,0 +1,241 @@
+    function verifyUnique(service) {
+      if (service in servicesPresent) {
+        throw new Meteor.Error(400, "Cannot merge accounts with duplicate service: " + service);
+      }
+      servicesPresent[service] = true;
+    }
+    winningUser.identities.forEach(function(identity) {
+      verifyUnique(identity.service);
+    });
+    losingUser.identities.forEach(function(identity) {
+      verifyUnique(identity.service);
+    });
+
+    // assert that there are no other pending merges?
+    Meteor.users.update(losingUser._id,
+                        {$set: {merging: {destinationUserId: winningUser._id, status: "pending",
+                                          sourceIdentities: losingUser.identities,
+                                          sourceServices: losingUser.services}},
+                         $unset: {identities: 1, services: 1}});
+
+  },
+
+  unmergeToAccount: function (destUserId) {
+    throw new Meteor.Error(400, "Accounts unmerging turned off for now. Needs more testing.");
+
+    check(destUserId, String);
+    if (!this.userId) {
+      throw new Meteor.Error(403, "Cannot merge accounts if not logged in.");
+    }
+    var sourceUser = Meteor.user();
+    if (!sourceUser.mergedUsers || sourceUser.mergedUsers.indexOf(destUserId) == -1) {
+      throw new Meteor.Error(403, "Current user was never merged with user " + destUserId);
+    }
+
+    // Now look up the destination user to figure out which identities and services
+    // to restore.
+
+    var destUser = Meteor.users.findOne(destUserId);
+
+    /// what if the status is still pending? Then we abort.
+    var oldServices = destUser.merging && destUser.merging.status === "done" &&
+        destUser.merging.sourceServices;
+
+    if (!oldServices) {
+      throw new Meteor.Error(400, "Cannot unmerge");
+    }
+    var oldIdentities = {};
+    var servicesToRemove = {};
+    destUser.merging.sourceIdentities.forEach(function(identity) {
+      oldIdentities[identity.id] = identity;
+      servicesToRemove["services." + identity.service] = 1;
+    });
+
+    var sourceIdentities = sourceUser.identities.filter(function (identity) {
+      return identity.id in oldIdentities;
+    });
+
+    // TODO what about the case when there are multiple email identities
+    // in a single emailToken service?
+
+    // what about quota, login?
+    Meteor.users.update(sourceUser._id,
+                        {$set: {unmerging: {destinationUserId: destUserId, status: "pending",
+                                            sourceIdentities: sourceIdentities,
+                                            sourceServices: oldServices}},
+                         $unset: servicesToRemove,
+                         $pull: {identities: {id: {$in: Object.keys(oldIdentities)}},
+                                 mergedUsers: destUserId} });
+  }
+});
+
+SandstormAccountsMerge.registerObservers = function (db, backend) {
+
+  Meteor.users.find({"merging.status": "pending"}).observe({
```

Comment:
This is pretty great in that if the server dies mid-merge it will retry on next start.

However, the function is currently not very robust against exceptions thrown within it. It would probably be a good idea to add a lot of try/catch blocks, not just around the whole thing but also individual parts, like maybe transferring of each grain. So if one grain fails to transfer, we don't get left in a totally inconsistent state -- only that one grain is affected. We can then inform the user of this and maybe fire off an alert email to ourselves (on Oasis at least).